### ISSUE TITLE
Omit `X-Powered-By` HTTP response header to avoid exposing Ratchet version to potential attackers

### ISSUE THREAD

**Issue by clue** [ratchetphp/Ratchet#999]
Right now, Ratchet always includes an `X-Powered-By` HTTP response header exposing the specific Ratchet version. This could potentially be used during reconnaissance to gain more information to prepare following attacks (see #932 and others).

At the moment, there does not appear to be a way to configure Ratchet to control this behavior. I would like to use this ticket as an opportunity to see if the default behavior makes sense, if we should remove this header in its entirety, if we should only include the major version or if we really need an option to control this.

For example, nginx always sends the `Server: nginx` header with an option to exclude the specific version (`server_tokens off`). For ReactPHP, we always send the `Server: ReactPHP/1` header with only the major version included (https://github.com/reactphp/http/pull/374). PHP always sends the `X-Powered-By: PHP/6.0.0` header unless [`expose_php`](https://www.php.net/manual/en/ini.core.php#ini.expose-php) is off.

In the meantime, you can always avoid this header by configuring your reverse proxy to remove this header. For nginx this should be as easy as:

```
proxy_hide_header X-Powered-By;
```

Note that irrespective of this header, the fact that Ratchet is serving the HTTP request can always be determined through other means, such as specific HTTP headers, timing attacks and additional implementation quirks. I do not currently see that this HTTP header alone constitutes a security issue on its own, but I definitely agree that this could contribute to seeing more targeted attacks should any security issues arise.

Any input is welcome :+1:

**Comment by SimonFrings** [comment]
> This is definitely the best solution, although it is radical. I can understand that you might not agree to use such a solution. But I think that this HTTP header does not add anything useful to Ratchet. Indeed, it is more likely to bring harms than benefits.

@Rezyan I agree with you, not sure tho how many users will be affected by this change (BC-break). This is something we want to evaluate here, but I think finding a temporary solution and completely removing it later may be the right way to go about this.

> If this HTTP header is intended to remain in the code, perhaps we could add a section in the README.md file that would indicate how to remove the HTTP header, both on Nginx and Apache. What do you think about this?

I think adding this as part of a documentation should be at least one solution here, completely agreed :+1:

**Comment by NkazimuloMvundla** [comment]
@Rezyan , have you come right with this? was planning on using Ratchet but if it has this sort of issue might look away

**Comment by SimonFrings** [comment]
@NkazimuloMvundla There are a few options in @clue's description of this ticket, for example you can also avoid this header by configuring your reverse proxy to remove this header. If you're using nginx, this looks something like this:

```
proxy_hide_header X-Powered-By;
```

If you have any input on how we should handle the situation with the `X-Powered-By` header, we're happy about your input and your reasoning on this :+1: 

FYI: I will mark these last answers as off-topic (mine included) as we didn't really progress in finding a good solution for the overall topic.

**Comment by ajnozari** [comment]
I feel like this should be configurable just as it is with PHP and whatever reverse proxy you use.

In the end it's a small thing and is security through obscurity but I still prefer it to be hidden.  For now we've updated our nginx config but I feel that having the ability to configure this within ratchet would be preferable than having to rely on the config of another part of our stack.

**Comment by PaulRotmann** [comment]
We're giving Ratchet some much-needed love and attention! As part of our issue cleanup initiative (#1100), we're reviewing security-related concerns like this one.

The concern about version disclosure is valid from a security hardening perspective. While the risk is generally low, we understand that some environments prefer to minimize information disclosure.

However, implementing a configuration option for this is more complex than it might appear, as the header is sent deep within the dependency chain across multiple abstraction layers. This would require significant architectural changes to expose the configuration option properly.

For immediate mitigation, you can configure your reverse proxy (nginx, Apache, etc.) to strip or modify the `Server` header before it reaches clients. This is actually a common practice for security hardening.

We're considering this enhancement for future versions, which would make such configuration options more feasible to implement.

We'll keep this issue open to track this security enhancement request.

If you want to support Ratchet, please consider [sponsoring](https://github.com/sponsors/reactphp) our work! ❤️

**Comment by DanielRuf** [comment]
Alternatively it should be possible to change the relevant file with composer-patches in projects.
